File: scala/src/test/scala/tddmicroexercises/tirepressuremonitoringsystem/TestAlarm.scala
package tddmicroexercises.tirepressuremonitoringsystem

import org.scalatest._
import org.scalatest.mock.MockitoSugar
import org.mockito.Mockito._

class TestAlarm extends FlatSpec with Matchers with MockitoSugar {

  def fixture =
    new {
      val sensor: Sensor = mock[Sensor]
      val alarm = new ExtendedAlarm(sensor);
    }

  it should "raise alarm on low pressure" in {
    val f = fixture
    when(f.sensor.popNextPressurePsiValue()).thenReturn(16.99)

    f.alarm.check()

    f.alarm.isAlarmOn() shouldBe true
    verify(f.sensor).popNextPressurePsiValue()
  }

  it should "raise alarm on high pressure" in {
    val f = fixture
    when(f.sensor.popNextPressurePsiValue()).thenReturn(21.01)

    f.alarm.check()

    f.alarm.isAlarmOn() shouldBe true
    verify(f.sensor).popNextPressurePsiValue()
  }

  it should "not raise alarm on normal pressure" in {
    val f = fixture
    when(f.sensor.popNextPressurePsiValue()).thenReturn(21.00)

    f.alarm.check()

    f.alarm.isAlarmOn() shouldBe false
    verify(f.sensor).popNextPressurePsiValue()
  }

}

File: Java/TirePressureMonitoringSystem/src/test/java/tddmicroexercises/tirepressuremonitoringsystem/TestAlarm.java
package tddmicroexercises.tirepressuremonitoringsystem;


import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class TestAlarm {

    private Sensor sensor;
    private Alarm alarm;

    @BeforeEach
    public void initialize() {
        sensor = mock(Sensor.class);
        alarm = new Alarm(sensor);
    }

    @ParameterizedTest
    @CsvSource({ "16.9, true",
            "21.1, true",
            "17, false",
            "21, false",
            "18, false",
            "19.2, false" })
    public void alarm_raises_when_pressure_outside_threshold(double pressure, boolean alarmOn) {
        when(sensor.popNextPressurePsiValue()).thenReturn(pressure);

        alarm.check();

        assertEquals(alarmOn, alarm.isAlarmOn());
    }

}
